### PR TITLE
Add algorithm to load gauss cube files

### LINKED ISSUE #35702: Add loader for gaussian cube format
**Is your feature request related to a problem? Please describe.**
SXD scientists want to load volumetric data (e.g. equivalent of mantid MDHisto workspaces) in the gaussian cube format.

**Describe the solution you'd like**
Add algorithms to save MD workspaces and load gauss cube format files as MD workspaces

**Describe alternatives you've considered**
None

**Additional context**
This is apparently a very common file format for this type of data

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Report to:** Matthias Gutmann (SXD)

**Summary of work**
Loader for Gaussian cube format (file format used to store volumetric data) - but data into a 3d MDHisto workspace.
This file format is used by SXD to store results of diffuse scattering simulations

**To test:**

(1) Run this script (same as unit test - so make sure you build test data)
```
ws = LoadGaussCube(Filename="gauss_cube_example.cube", Names=['[H,0,0]','[0,K,0]','[0,0,L]'],
                          Frames='HKL,HKL,HKL',
                          Units='rlu,rlu,rlu')
```
(2) Check you get an error if there are not 3 elements in the `Names`, `Frames` and `Units` input parameters

Fixes #35702 

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/LoadGaussCube.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// SPDX - License - Identifier: GPL - 3.0 +
+#pragma once
+
+#include "MantidAPI/DataProcessorAlgorithm.h"
+#include "MantidMDAlgorithms/DllConfig.h"
+
+namespace Mantid {
+namespace MDAlgorithms {
+
+/** LoadGaussCube : TODO: DESCRIPTION
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/inc/MantidMDAlgorithms/LoadGaussCube.h:15]: Did you mean to leave this to do in?

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+/// Algorithm's version for identification. @see Algorithm::version
+int LoadGaussCube::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string LoadGaussCube::category() const { return "MDAlgorithms"; }
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string LoadGaussCube::summary() const {
+  return "Algorithm to load gauss cube files and output an 3D MDHistoWorkspace.";
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/src/LoadGaussCube.cpp:40]:
```suggestion
  return "Algorithm to load gauss cube files and output a 3D MDHistoWorkspace.";
```

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void LoadGaussCube::exec() {
+  std::string Filename = getProperty("Filename");
+
+  // open file
+  std::ifstream in(Filename.c_str());
+  // skip header
+  for (int iline = 0; iline < 2; ++iline) {
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/src/LoadGaussCube.cpp:87]: what is `iline`?

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    in.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+  }
+
+  // read lower extent
+  std::vector<double> extents(6, 0.0);
+  getWord(in, false); // ignore the first element of this row
+  std::string str;
+  for (size_t idim = 0; idim < 3; ++idim) {
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/src/LoadGaussCube.cpp:95]: idim?

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/LoadGaussCube.h (hunk 1 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/inc/MantidMDAlgorithms/LoadGaussCube.h:15]: Nope, will remove!

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp (hunk 3 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/src/LoadGaussCube.cpp:87]: index of line

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp (hunk 4 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/src/LoadGaussCube.cpp:95]: index of dimension

File: Framework/MDAlgorithms/test/LoadGaussCubeTest.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+public:
+  static LoadGaussCubeTest *createSuite() { return new LoadGaussCubeTest(); }
+  static void destroySuite(LoadGaussCubeTest *suite) { delete suite; }
+
+  void test_Init() {
+    LoadGaussCube alg;
+    TS_ASSERT_THROWS_NOTHING(alg.initialize())
+    TS_ASSERT(alg.isInitialized())
+  }
+
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/test/LoadGaussCubeTest.h:29]: can we have a test to check the validation?

File: docs/source/algorithms/LoadGaussCube-v1.rst [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+Usage
+-----
+
+**Example - LoadGaussCube**
+
+.. code-block:: python
+
+    from mantid.simpleapi import *
+
+    ws = LoadGaussCube(Filename="gauss_cube_example.cube", Names=['[H,0,0]','[0,K,0]','[0,0,L]'], Frames='HKL,HKL,HKL', Units='rlu,rlu,rlu')
```
Comment (AnthonyLim23) [inline docs/source/algorithms/LoadGaussCube-v1.rst:26]: can we not add a doc test here? If we print something we can say what the expected output is.

File: Framework/MDAlgorithms/src/LoadGaussCube.cpp (hunk 3 above)
Comment (AnthonyLim23) [reply Framework/MDAlgorithms/src/LoadGaussCube.cpp:87]: I would expect it to be either `indexOfLine` or `line_i` (I personally like the second one better, but thats me with my science hat on)

File: Framework/MDAlgorithms/test/LoadGaussCubeTest.h (hunk 5 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/test/LoadGaussCubeTest.h:29]: Yep good shout!

File: docs/source/algorithms/LoadGaussCube-v1.rst (hunk 6 above)
Comment (RichardWaiteSTFC) [reply docs/source/algorithms/LoadGaussCube-v1.rst:26]: Can do, but given the output is a single workspace, I'm not sure what that would add over the unit tests?

Comment (AnthonyLim23) [reply docs/source/algorithms/LoadGaussCube-v1.rst:26]: Just an extra layer of testing for us. It also tells users what sort of result to expect, you can use it to highlight key differences to "normal" output. I always feel happier seeing a test in the docs as I can check if it works as the docs expect

**Comment by AnthonyLim23** [comment]
I cannot use it from the algorithm dialog 
image

**Comment by RichardWaiteSTFC** [comment]
> I cannot use it from the algorithm dialog image

As discussed offline this is due to the way strings of lists are parsed. In this case we want to ignore the `,` in e.g. `[H,0,0]` but not the `,` that delimits the different axes names - to do this in a script you put the names in a python list, but you cannot put it in a single string as required in the algorithm dialog (note you need to remove the speech quotes as it is already interpreted as a string).
It was a known issue with `CreateMDHistoWorkspace` but there is a workaround for this use-case that I can add try to use in this PR
https://github.com/mantidproject/mantid/blob/c7d4d55d719739e9e2f029ee88c62f9991f2d577/Framework/MDAlgorithms/src/ImportMDHistoWorkspaceBase.cpp#L203

**Review by AnthonyLim23 — APPROVED** [review]
It works as expected, even from the algorithm dialog. The code is also very readable now. :shipit:

**Pull request merged**